src/Castle.Windsor/MicroKernel/Arguments.cs

```diff
@@ -19,208 +19,240 @@ namespace Castle.MicroKernel
+		/// </summary>
+		public Arguments(IDictionary values)
+			: this()
 		{
-			if (customComparers == null || customComparers.Length == 0)
-			{
-				arguments = new Dictionary<object, object>(new ArgumentsComparer());
-			}
-			else
+			foreach (DictionaryEntry entry in values)
 			{
-				arguments = new Dictionary<object, object>(new ArgumentsComparerExtended(customComparers));
+				Add(entry.Key, entry.Value);
 			}
 		}
 
-		public int Count
-		{
-			get { return arguments.Count; }
-		}
-
+		/// <summary>
+		/// Indexer for creating named/typed dependency arguments for <see cref="IWindsorContainer.Resolve{T}(Castle.MicroKernel.Arguments)"/>
+		/// </summary>
 		public object this[object key]
 		{
-			get { return arguments[key]; }
-			set { arguments[key] = value; }
+			get => dictionary[key];
+			set
+			{
+				EnsureWritable();
+				dictionary[key] = value;
+			}
 		}
 
-		public ICollection Keys
-		{
-			get { return arguments.Keys; }
-		}
+		public int Count => dictionary.Count;
 
-		public ICollection Values
-		{
-			get { return arguments.Values; }
-		}
+		public ICollection Keys => dictionary.Keys;
 
-		bool ICollection.IsSynchronized
-		{
-			get { return arguments.IsSynchronized; }
-		}
+		public ICollection Values => dictionary.Values;
+
+		bool ICollection.IsSynchronized => dictionary.IsSynchronized;
+
+		object ICollection.SyncRoot => dictionary.SyncRoot;
+
+		bool IDictionary.IsFixedSize => dictionary.IsFixedSize;
+
+		bool IDictionary.IsReadOnly => isReadOnly;
+
+		public static readonly Arguments Empty = new Arguments { isReadOnly = true };
 
-		object ICollection.SyncRoot
+		public void Add(object key, object value)
 		{
-			get { return arguments.SyncRoot; }
+			EnsureWritable();
+			dictionary.Add(key, value);
 		}
 
-		bool IDictionary.IsFixedSize
+		public void Clear()
 		{
-			get { return arguments.IsFixedSize; }
+			EnsureWritable();
+			dictionary.Clear();
 		}
 
-		bool IDictionary.IsReadOnly
+		public Arguments Clone()
 		{
-			get { return arguments.IsReadOnly; }
+			return new Arguments(dictionary);
 		}
 
-		public void Add(object key, object value)
+		void ICollection.CopyTo(Array array, int index)
 		{
-			arguments.Add(key, value);
+			dictionary.CopyTo(array, index);
 		}
 
-		public void Clear()
+		public bool Contains(object key)
 		{
-			arguments.Clear();
+			return dictionary.Contains(key);
 		}
 
-		public bool Contains(object key)
+		IEnumerator IEnumerable.GetEnumerator()
 		{
-			return arguments.Contains(key);
+			return GetEnumerator();
 		}
 
 		public IDictionaryEnumerator GetEnumerator()
 		{
-			return arguments.GetEnumerator();
+			return dictionary.GetEnumerator();
 		}
 
-		public void Remove(object key)
+		/// <summary>
+		/// Inserts a named argument. If the argument already exists it will be overwritten.
+		/// </summary>
+		public virtual Arguments Insert(string key, object value)
```

Comment:
Glad to know you agree with that comment, I really thought I might have been going over the top.

Yes, the class isn't meant to have special features, pretty much a dumb dictionary with some useful methods to get key/value pairs into it from different sources. Removed the virtual modifier, I'd actually be inclined to seal the class but that is probably too far 😉.

Reply:
> I'd actually be inclined to seal the class but that is probably too far

I mean, not as far as I'm concerned. :D If a class isn't designed with inheritance in mind, you could treat it a bugfix to add `sealed`. (But only if you expect that no one would have inherited from it, or if you like the idea of causing them to rewrite if so.)

Reply:
That, and it is easier to un-seal a class later on than to seal it.

Reply:
I think classes should have been sealed by default in C# and that it's sign to be suspicious of when a class is not either `sealed` or `abstract`. This is opinionated and not everyone agrees with it, so take it with a grain of salt. From a lib design standpoint, starting sealed is safest as @stakx points out, and sealing it guides users right away on how they are going to be interacting with the class.

Reply:
If a user wants to extend this class with another utility function they can always do that with an extension method on `Arguments` (and call `Add`), a subclass couldn't as easily surface new methods because Windsor exposes this object as type `Arguments` not your subclass anyway, and the internal dictionary is `private readonly` not `protected`.

Castle used to always make every class public and everything extensible because everyone used to just run off the subversion trunk, backwards compatibility wasn't that important, this probably stemmed from the crowd coming from Java where things are virtual by default. However over the last few years we've been changing quite a bit of that in Castle Core because things that weren't designed as extension points or really for public use has been used which has tied the implementation into the public API.

I've sealed this class for now until we have a good use case for the alternative. Many thanks guys for the discussion.